internal/services/storagemover/storage_mover_source_endpoint_resource.go

```diff
@@ -0,0 +1,257 @@
+				string(endpoints.NfsVersionNFSvThree),
+			}, false),
+		},
+
+		"description": {
+			Type:         pluginsdk.TypeString,
+			Optional:     true,
+			ValidateFunc: validation.StringIsNotEmpty,
+		},
+	}
+}
+
+func (r StorageMoverSourceEndpointResource) Attributes() map[string]*pluginsdk.Schema {
+	return map[string]*pluginsdk.Schema{}
+}
+
+func (r StorageMoverSourceEndpointResource) Create() sdk.ResourceFunc {
+	return sdk.ResourceFunc{
+		Timeout: 30 * time.Minute,
+		Func: func(ctx context.Context, metadata sdk.ResourceMetaData) error {
+			var model StorageMoverSourceEndpointModel
+			if err := metadata.Decode(&model); err != nil {
+				return fmt.Errorf("decoding: %+v", err)
+			}
+
+			client := metadata.Client.StorageMover.EndpointsClient
+			storageMoverId, err := storagemovers.ParseStorageMoverID(model.StorageMoverId)
+			if err != nil {
+				return err
+			}
+
+			id := endpoints.NewEndpointID(storageMoverId.SubscriptionId, storageMoverId.ResourceGroupName, storageMoverId.StorageMoverName, model.Name)
+			existing, err := client.Get(ctx, id)
+			if err != nil && !response.WasNotFound(existing.HttpResponse) {
+				return fmt.Errorf("checking for existing %s: %+v", id, err)
+			}
+
+			if !response.WasNotFound(existing.HttpResponse) {
+				return metadata.ResourceRequiresImport(r.ResourceType(), id)
+			}
+
+			properties := endpoints.Endpoint{
+				Name: utils.String(model.Name),
+				Properties: endpoints.NfsMountEndpointProperties{
+					Export:     model.Export,
+					Host:       model.Host,
+					NfsVersion: &model.NfsVersion,
+				},
+				Type: utils.String(string(endpoints.EndpointTypeNfsMount)),
+			}
+
+			if model.Description != "" {
+				nfsProperties := properties.Properties.(endpoints.NfsMountEndpointProperties)
+				nfsProperties.Description = utils.String(model.Description)
+				properties.Properties = nfsProperties
+			}
+
+			if _, err := client.CreateOrUpdate(ctx, id, properties); err != nil {
+				return fmt.Errorf("creating %s: %+v", id, err)
+			}
+
+			metadata.SetID(id)
+			return nil
+		},
+	}
+}
+
+func (r StorageMoverSourceEndpointResource) Update() sdk.ResourceFunc {
+	return sdk.ResourceFunc{
+		Timeout: 30 * time.Minute,
+		Func: func(ctx context.Context, metadata sdk.ResourceMetaData) error {
+			client := metadata.Client.StorageMover.EndpointsClient
+
+			id, err := endpoints.ParseEndpointID(metadata.ResourceData.Id())
+			if err != nil {
+				return err
+			}
+
+			var model StorageMoverSourceEndpointModel
+			if err := metadata.Decode(&model); err != nil {
+				return fmt.Errorf("decoding: %+v", err)
+			}
+
+			resp, err := client.Get(ctx, *id)
+			if err != nil {
+				return fmt.Errorf("retrieving %s: %+v", *id, err)
+			}
+
+			properties := resp.Model
+			if properties == nil {
+				return fmt.Errorf("retrieving %s: model was nil", *id)
+			}
+
+			if metadata.ResourceData.HasChange("description") {
+				nfsProperties := properties.Properties.(endpoints.NfsMountEndpointProperties)
```

Comment:
we're not type checking this, if the cast fails this'll panic

Reply:
Fixed.